crates/xhr/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+    /// documentation.
+    #[derive(Debug, Clone)]
+    pub struct XmlHttpRequest {
+        xhr: web_sys::XmlHttpRequest,
+    }
+
+    // This is so it can be used as an event target by gloo-events.
+    impl AsRef<web_sys::EventTarget> for XmlHttpRequest {
+        fn as_ref(&self) -> &web_sys::EventTarget {
+            self.xhr.as_ref()
+        }
+    }
+
+    impl XmlHttpRequest {
+        /// Initialize an XmlHttpRequest. To actually perform the request, you will need to
+        /// call at least `open()` and `send()`.
+        pub fn new() -> Self {
+            // we assume this is safe because all browsers that support webassembly
+            // implement XmlHttpRequest.
+            let xhr = web_sys::XmlHttpRequest::new().unwrap_throw();
+            XmlHttpRequest { xhr }
+        }
+
+        /// Access the underlying raw `web_sys::XmlHttpRequest`.
+        pub fn as_raw(&self) -> &web_sys::XmlHttpRequest {
+            &self.xhr
+        }
+
+        /// Returns an event target for upload events.
+        ///
+        /// See the [module docs](crate::callback) and `XmlHttpRequestUpload` for more information.
+        pub fn upload(&self) -> XmlHttpRequestUpload {
+            XmlHttpRequestUpload {
+                upload: self.xhr.upload().unwrap_throw(),
+            }
+        }
+
+        /// The `XMLHttpRequest` method `open()` initializes a newly-created request, or
+        /// re-initializes an existing one.
+        ///
+        /// [MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/open)
+        pub fn open(&self, method: &http::Method, url: &str) {
+            self.xhr.open(&method.to_string(), url).unwrap_throw()
+        }
+
+        /// Abort the request.
+        ///
+        /// This will fire an `abort` event on the `XmlHttpRequest`.
+        pub fn abort(&self) {
+            self.xhr.abort().unwrap_throw()
+        }
+
+        /// Perform the request.
+        ///
+        /// Thise returns immediately - attach events handlers to be notified of the
+        /// progress and outcome of the request.
+        pub fn send<B: XhrBody>(&self, body: SendBody<B>) {
+            match body {
+                SendBody::None => self.send_without_body(),
+                SendBody::Body(body) => self.send_with_body(body),
+            }
+        }
+
+        /// `send()` without a body.
+        pub fn send_without_body(&self) {
+            self.xhr
+                .send()
+                .expect_throw("Error sending request. Did you forget to call `open`?")
+        }
+
+        /// Actually send the request. In order to know the outcome, you have to attach
+        /// `load`, `abort` and `error` event listeners.
+        pub fn send_with_body<B: XhrBody>(&self, body: B) {
+            body.send(&self.xhr)
+                .expect_throw("Error sending request. Did you forget to call `open`?")
```

Comment:
I would've expected all the `send_*` methods to return `Result` instead of panicking. Could you expand on the reasoning here why panicking is preferred over `Result`?

Reply:
I went the shortest route for error handling because there was no clear prior art/guideline there, but I agree that in principle a `Result` with a type error variant is ideal.